include/xsimd/xsimd.hpp

```diff
@@ -12,7 +12,15 @@
 #ifndef XSIMD_HPP
 #define XSIMD_HPP
 
-#if defined(__GNUC__) || defined(__clang__)
+#if defined(__has_cpp_attribute)
+// if this is true, then the compiler supports feature test macros
+#if __has_cpp_attribute(nodiscard) >= 201603L
+// if this is true, then the compiler supports [[nodiscard]] without a message
+#define XSIMD_NO_DISCARD [[nodiscard]]
+#endif
+#elif __cplusplus >= 201703L // this means we are using C++17 or higher
+#define XSIMD_NO_DISCARD [[nodiscard]]
```

Comment:
Do we need to filter on the date? Why is just `#if __has_cpp_attribute(nodiscard` not enough?

Reply:
The reason is that these [feature test macros](https://en.cppreference.com/w/User:D41D8CD98F/feature_testing_macros) were actually added in C++20, whereas `[[nodiscard]]` has been in the language since C++17.